File: sdk/php/src/Internal/Rest/DefaultKucoinRestAPIImpl.php
<?php

namespace KuCoin\UniversalSDK\Internal\Rest;

use Exception;
use KuCoin\UniversalSDK\Api\KucoinRestService;
use KuCoin\UniversalSDK\Generate\Service\AccountService;
use KuCoin\UniversalSDK\Generate\Service\AccountServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\AffiliateService;
use KuCoin\UniversalSDK\Generate\Service\AffiliateServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\BrokerService;
use KuCoin\UniversalSDK\Generate\Service\BrokerServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\CopyTradingService;
use KuCoin\UniversalSDK\Generate\Service\CopyTradingServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\EarnService;
use KuCoin\UniversalSDK\Generate\Service\EarnServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\FuturesService;
use KuCoin\UniversalSDK\Generate\Service\FuturesServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\MarginService;
use KuCoin\UniversalSDK\Generate\Service\MarginServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\SpotService;
use KuCoin\UniversalSDK\Generate\Service\SpotServiceImpl;
use KuCoin\UniversalSDK\Generate\Service\VIPLendingService;
use KuCoin\UniversalSDK\Generate\Service\VIPLendingServiceImpl;
use KuCoin\UniversalSDK\Generate\Version;
use KuCoin\UniversalSDK\Internal\Infra\DefaultTransport;
use KuCoin\UniversalSDK\Model\ClientOption;

class DefaultKucoinRestAPIImpl implements KucoinRestService
{
    private $accountService;
    private $affiliateService;
    private $brokerService;
    private $copyTradingService;
    private $earnService;
    private $futuresService;
    private $marginService;
    private $spotService;
    private $vipLendingService;

    public function __construct(ClientOption $option)
    {
        if ($option->transportOption == null) {
            throw new Exception("no transport option provided");
        }
        $transport = new DefaultTransport($option, Version::SDK_VERSION);
        $this->accountService = new AccountServiceImpl($transport);
        $this->affiliateService = new AffiliateServiceImpl($transport);
        $this->brokerService = new BrokerServiceImpl($transport);
        $this->copyTradingService = new CopyTradingServiceImpl($transport);
        $this->earnService = new EarnServiceImpl($transport);
        $this->futuresService = new FuturesServiceImpl($transport);
        $this->marginService = new MarginServiceImpl($transport);
        $this->spotService = new SpotServiceImpl($transport);
        $this->vipLendingService = new VIPLendingServiceImpl($transport);
    }

    public function getAccountService(): AccountService
    {
        return $this->accountService;
    }

    public function getAffiliateService(): AffiliateService
    {
        return $this->affiliateService;
    }

    public function getBrokerService(): BrokerService
    {
        return $this->brokerService;
    }

    public function getCopytradingService(): CopyTradingService
    {
        return $this->copyTradingService;
    }

    public function getEarnService(): EarnService
    {
        return $this->earnService;
    }

    public function getFuturesService(): FuturesService
    {
        return $this->futuresService;
    }

    public function getMarginService(): MarginService
    {
        return $this->marginService;
    }

    public function getSpotService(): SpotService
    {
        return $this->spotService;
    }

    public function getVipLendingService(): VIPLendingService
    {
        return $this->vipLendingService;
    }
}

File: sdk/java/src/main/java/com/kucoin/universal/sdk/internal/rest/DefaultKucoinRestAPIImpl.java
package com.kucoin.universal.sdk.internal.rest;

import com.kucoin.universal.sdk.api.KucoinRestService;
import com.kucoin.universal.sdk.generate.Version;
import com.kucoin.universal.sdk.generate.service.AccountService;
import com.kucoin.universal.sdk.generate.service.AccountServiceImpl;
import com.kucoin.universal.sdk.generate.service.AffiliateService;
import com.kucoin.universal.sdk.generate.service.AffiliateServiceImpl;
import com.kucoin.universal.sdk.generate.service.BrokerService;
import com.kucoin.universal.sdk.generate.service.BrokerServiceImpl;
import com.kucoin.universal.sdk.generate.service.CopyTradingService;
import com.kucoin.universal.sdk.generate.service.CopyTradingServiceImpl;
import com.kucoin.universal.sdk.generate.service.EarnService;
import com.kucoin.universal.sdk.generate.service.EarnServiceImpl;
import com.kucoin.universal.sdk.generate.service.FuturesService;
import com.kucoin.universal.sdk.generate.service.FuturesServiceImpl;
import com.kucoin.universal.sdk.generate.service.MarginService;
import com.kucoin.universal.sdk.generate.service.MarginServiceImpl;
import com.kucoin.universal.sdk.generate.service.SpotService;
import com.kucoin.universal.sdk.generate.service.SpotServiceImpl;
import com.kucoin.universal.sdk.generate.service.VIPLendingService;
import com.kucoin.universal.sdk.generate.service.VIPLendingServiceImpl;
import com.kucoin.universal.sdk.internal.infra.DefaultTransport;
import com.kucoin.universal.sdk.internal.interfaces.Transport;
import com.kucoin.universal.sdk.model.ClientOption;
import lombok.extern.slf4j.Slf4j;

/**
 * DefaultKucoinRestAPIImpl is the default implementation of {@link KucoinRestService}. It wires
 * every generated REST service implementation with a shared {@link DefaultTransport}.
 */
@Slf4j
public final class DefaultKucoinRestAPIImpl implements KucoinRestService {

  private final AccountService accountService;
  private final AffiliateService affiliateService;
  private final BrokerService brokerService;
  private final CopyTradingService copyTradingService;
  private final EarnService earnService;
  private final FuturesService futuresService;
  private final MarginService marginService;
  private final SpotService spotService;
  private final VIPLendingService vipLendingService;
  private final Transport transport;

  public DefaultKucoinRestAPIImpl(ClientOption option) {
    if (option.getTransportOption() == null) {
      throw new RuntimeException("no transport option provided");
    }
    transport = new DefaultTransport(option, Version.SDK_VERSION);

    this.accountService = new AccountServiceImpl(transport);
    this.affiliateService = new AffiliateServiceImpl(transport);
    this.brokerService = new BrokerServiceImpl(transport);
    this.copyTradingService = new CopyTradingServiceImpl(transport);
    this.earnService = new EarnServiceImpl(transport);
    this.futuresService = new FuturesServiceImpl(transport);
    this.marginService = new MarginServiceImpl(transport);
    this.spotService = new SpotServiceImpl(transport);
    this.vipLendingService = new VIPLendingServiceImpl(transport);

    log.info("SDK version: {}", Version.SDK_VERSION);
  }

  @Override
  public AccountService getAccountService() {
    return accountService;
  }

  @Override
  public AffiliateService getAffiliateService() {
    return affiliateService;
  }

  @Override
  public BrokerService getBrokerService() {
    return brokerService;
  }

  @Override
  public CopyTradingService getCopytradingService() {
    return copyTradingService;
  }

  @Override
  public EarnService getEarnService() {
    return earnService;
  }

  @Override
  public FuturesService getFuturesService() {
    return futuresService;
  }

  @Override
  public MarginService getMarginService() {
    return marginService;
  }

  @Override
  public SpotService getSpotService() {
    return spotService;
  }

  @Override
  public VIPLendingService getVipLendingService() {
    return vipLendingService;
  }

  @Override
  public void closeService() {
    transport.close();
  }
}
